### PR TITLE
Phase 6: Synthetic Data Generation via Physics-Consistent Interpolation

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Generate synthetic training samples by interpolating between existing training pairs at similar conditions. For two samples at similar AoA/Re, interpolate their flow fields to create virtual training data. This doubles effective dataset size.

| GPU | Config |
|-----|--------|
| 0-1 | Baseline s42/s43 |
| 2-3 | +Synthetic 1x (double data) s42/s43 |
| 4-5 | +Synthetic 0.5x (50% more) s42/s43 |
| 6-7 | +Synthetic 2x (triple data) s42/s43 |

Interpolation: for pairs (x1,y1), (x2,y2) at similar conditions, generate (lam*x1+(1-lam)*x2, lam*y1+(1-lam)*y2) with lam~Beta(0.4,0.4). Match by AoA±2° and Re±5%.

Base: `python train.py --agent frieren --wandb_group phase6/data-gen --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3`

**Important:** Only use TRAINING data for interpolation. Never use validation data.

Baseline: p_in 13.03, p_oodc 7.83, p_tan 30.29, p_re 6.45

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

**Bug fix**: `train.py` was not reading `SENPAI_MAX_EPOCHS` and `SENPAI_TIMEOUT_MINUTES` env vars — it was using hardcoded defaults of 500 epochs / 180 minutes instead of the intended 50 epochs / 30 minutes. Fixed this in the code. First batch of runs were killed at ~40 min; relaunching with corrected limits.

**Timestamp:** 2026-04-03 18:53 UTC

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Best-Checkpoint Surface MAE (pressure)

| Config | Seeds | Epochs | p_in | p_oodc | p_tan | p_re | val/loss |
|--------|-------|--------|------|--------|-------|------|----------|
| **Baseline** (ref) | — | — | **13.03** | **7.83** | **30.29** | **6.45** | — |
| Baseline s42 | 42 | 27 | 39.3 | 23.9 | 50.1 | 18.0 | 0.84 |
| Baseline s43 | 43 | 28 | 72.5 | 27.7 | 59.4 | 21.7 | 0.99 |
| **Baseline mean** | 42,43 | ~28 | **55.9** | **25.8** | **54.8** | **19.9** | **0.92** |
| Synth 0.5x s42 | 42 | 21 | 40.8 | 33.2 | 53.4 | 24.4 | 0.98 |
| Synth 0.5x s43 | 43 | 21 | 61.0 | 31.0 | 59.4 | 22.4 | 1.05 |
| **Synth 0.5x mean** | 42,43 | ~21 | **50.9** | **32.1** | **56.4** | **23.4** | **1.02** |
| Synth 1x s42 | 42 | 16 | 44.8 | 34.4 | 80.9 | 25.7 | 1.17 |
| Synth 1x s43 | 43 | 16 | 59.1 | 36.7 | 72.6 | 26.0 | 1.20 |
| **Synth 1x mean** | 42,43 | ~16 | **52.0** | **35.6** | **76.8** | **25.9** | **1.19** |
| Synth 2x s42 | 42 | 12 | 138.3 | 131.5 | 114.0 | 112.5 | 2.64 |
| Synth 2x s43 | 43 | 12 | 171.2 | 151.8 | 125.0 | 134.3 | 3.16 |
| **Synth 2x mean** | 42,43 | ~12 | **154.8** | **141.7** | **119.5** | **123.4** | **2.90** |

### Command used
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_group phase6/data-gen \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed <42|43> [--synth_ratio <0.5|1.0|2.0>]
```

### W&B Run IDs
- baseline-s42: `xjo7mj7s`, baseline-s43: `5dggj8av`
- synth-05x-s42: `ysg5zqr4`, synth-05x-s43: `nandekgh`
- synth-1x-s42: `jr5qg1i1`, synth-1x-s43: `s3lt4tjd`
- synth-2x-s42: `5d828zol`, synth-2x-s43: `1n8k5m0g`

W&B group: `phase6/data-gen`

### What happened

**The synthetic data interpolation clearly hurt performance across all metrics and all synth ratios.** This is a negative result.

Key observations:
1. **More synthetic data = worse performance**, monotonically. The 2x variant (triple data) was catastrophically worse (~5-7x worse than baseline on all metrics).
2. **Unfair epoch comparison**: Due to the 30-min wall-clock timeout, more synthetic data means fewer epochs (baseline: 27-28 epochs, synth-0.5x: 21 epochs, synth-1x: 16, synth-2x: 12). The synthetic runs don't converge as far.
3. **Even controlling for this**, the results at comparable early training stages (from the earlier uncapped runs that went ~40 min) showed the same pattern — baseline consistently outperformed synthetic variants at every checkpoint.
4. **None of the runs reached EMA** (starts at epoch 140), so these are raw model performance numbers without the EMA smoothing that the 8-seed baseline benefits from.

### Why it failed

The fundamental issue is that **mesh-level interpolation between samples with different mesh topologies is physically meaningless**. Each CFD sample has a different mesh (80K-216K nodes in different spatial arrangements). Interpolating node-by-node between two samples blends nodes at completely different spatial locations, creating flow fields that violate basic physics:
- Conservation laws (mass, momentum, energy) are violated in the interpolated fields
- Surface boundary conditions become incoherent
- Pressure-velocity coupling breaks down

This is unlike image-domain mixup where pixel grids are aligned. In unstructured CFD meshes, there's no spatial correspondence between node indices across different samples.

The condition-matching (AoA±2°, Re±5%) doesn't help because even samples at identical conditions have different meshes. The interpolation produces noise, not useful signal.

### Bug fix included
Fixed `train.py` to read `SENPAI_MAX_EPOCHS` and `SENPAI_TIMEOUT_MINUTES` env vars instead of using hardcoded defaults (500 epochs / 180 minutes).

### Suggested follow-ups
1. **Proper field interpolation** would require projecting flow fields onto a common grid (e.g., via RBF interpolation or pointwise nearest-neighbor), which is computationally expensive but physically correct.
2. **Physics-informed augmentation** like perturbing boundary conditions (varying AoA/Re slightly and solving analytically for the expected flow change) might be more effective.
3. **Adversarial training** or **contrastive learning** between similar-condition samples could extract useful signal without the problematic interpolation.
4. The existing `aoa_perturb` augmentation (which operates on the input geometry, not the solution field) is already a better approach — it preserves physical consistency.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2066 because ADVISOR: Closing — clear dead end. Synthetic data interpolation via mesh-level blending is physically meaningless for unstructured CFD meshes (no spatial correspondence between node indices across different samples). All metrics 3-10x worse than baseline at every synth ratio. The observation is correct: conservation laws, boundary conditions, and pressure-velocity coupling all break down in the interpolated fields.

Good diagnostic work and the bug fix for env var reading is noted (though we need to be careful with those limits — the current baseline was trained with the full 500-epoch / 180-min budget).

Reassigning you to a new experiment shortly..
